README
This is Base4R, a Ruby client library for the Google Base API.

All code is copyright 2007-2008 Dan Dukeson.

This code is licenced under the GNU General Public License, which can be found in the LICENCE file.

This release is quite basic and doesn't attempt to cover the whole of the Google Base API. In particular the code is write-only, i.e. there is functionality to create, update and delete Base items but nothing to query the Base API. Contributions of code to add this functionality would be welcomed.

You can contact the author at dandukeson at gmail dot com, or raise feature requests through the project page hosted at http://code.google.com/p/base4r/issues/list

To run the unit tests you should make a file test/test_config.rb with content like the following, but replacing the example values with your Google Base login details and API key.

TEST_CONFIG = {
  :username => 'user@example.com',
  :password => 'password',
  :api_key  => 'example_api_key'
}




CHANGES
v0.1 2008-06-01 
     First release.

v0.2 2008-08-17 
     Add a required "unique_id" attribute to Item, as this is now required for all Base items see http://base.google.com/support/bin/answer.py?answer=78170#id.
     Workaround xml formatting weirdness that has appeared since I last ran the unit tests by removing inter-tag whitespace in unit tests.
     Remove the default values for "target_country" and "item_lang" in the constructor for UniversalItem and ProductItem to make the code a little less Brit-centric.
     
     


test/item_test.rb
#
# tests for Item and children
#

require 'test/unit'
require 'item'

module Base4R

  class ItemTest < Test::Unit::TestCase #:nodoc:

    def setup
      
      @item = UniversalItem.new('ABC123',
                                'Dan',
                                'dan@example.com',
                                'A wonderful item with many featureful benefits',
                                'http://example.com/greatstuff/item/goo',
                                'Beneficial Product',
                                '0114 1234567',
                                'generic item',
                                'GB',
                                'EN')

      @expiration_time = Time.now+7*60*60*24

      @item.application = 'base4r'
      @item.expiration_date = @expiration_time
      @item.add_image_link('http://example.com/images/big1.png')
      @item.add_image_link('http://example.com/images/big2.png')
      @item.add_label('big')
      @item.add_label('clever')
      @item.add_label('funny')

    end

    
    def test_attributes_present
      assert_equal 14, @item.attributes.size
    end

    def test_xml

      expected_xml = "<?xml version='1.0'?><entry xmlns:g='http://base.google.com/ns/1.0' xmlns='http://www.w3.org/2005/Atom'><author><name>Dan</name><email>dan@example.com</email></author><category term='googlebase.item' scheme='http://www.google.com/type'/><title>Beneficial Product</title><link href='http://example.com/greatstuff/item/goo' rel='alternate' type='text/html'/><content>A wonderful item with many featureful benefits</content><g:contact_phone type='text'>0114 1234567</g:contact_phone><g:item_language type='text'>EN</g:item_language><g:item_type type='text'>generic item</g:item_type><g:target_country type='text'>GB</g:target_country><g:id type='text'>ABC123</g:id><g:application type='text'>base4r</g:application><g:expiration_date type='dateTime'>#{@expiration_time}</g:expiration_date><g:image_link>http://example.com/images/big1.png</g:image_link><g:image_link>http://example.com/images/big2.png</g:image_link><g:label type='text'>big</g:label><g:label type='text'>clever</g:label><g:label type='text'>funny</g:label></entry>"
      actual_xml = ""
      @item.to_xml.write(actual_xml)
      self.assert_equal expected_xml, actual_xml 
      
    end    
  end
end


test/attribute_test.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# tests for Attribute and subclasses
#
require 'test/unit'
require 'attribute'

module Base4R
  
  class AttributeTest < Test::Unit::TestCase #:nodoc:

    def setup
      
      @text_attr = TextAttribute.new('testname', 'testvalue', :g)
      @url_attr = UrlAttribute.new('testname', 'testvalue', :g)
      @date_time_attr = DateTimeAttribute.new('testname', 'testvalue', :g)
      @int_attr = IntAttribute.new('testname', 'testvalue', :g)

    end
    
    def test_type_names
      assert_equal 'text', @text_attr.type_name
      assert_equal 'url', @url_attr.type_name
      assert_equal 'dateTime', @date_time_attr.type_name
      assert_equal 'int', @int_attr.type_name
    end

    def test_accessors

      @text_attr.name = 'foo'
      @text_attr.value = 'bar'
      @text_attr.type_name = 'text'
      @text_attr.namespace = :atom
      
      assert_equal 'foo', @text_attr.name
      assert_equal 'bar', @text_attr.value
      assert_equal 'text',@text_attr.type_name
      assert_equal :atom, @text_attr.namespace
    end

  end

end


test/base_client_test.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#
# tests for BaseClient
#

require 'test/unit'
require 'base_client'
require 'net/http'
require 'test/test_config'

module Base4R

  class BaseClientTest < Test::Unit::TestCase #:nodoc:

    def setup      

      @username = TEST_CONFIG[:username]
      @password = TEST_CONFIG[:password]
      @api_key  = TEST_CONFIG[:api_key]
      
      @base_client = Base4R::BaseClient.new(@username, @password, @api_key, false)

      @item = UniversalItem.new('TEST123',
                                'base4r test author',
                               'base4r@example.com',
                               'This description describes a new item',
                               'http://example.com/items/dir/34545.html',
                               'Title of a new item',
                               '0114 1111111',
                               'products',
                                'GB',
                                'EN')

    end

    def teardown
    end

    def test_feed_path

      assert_equal '/base/feeds/items/', @base_client.feed_path     
      another_client = BaseClient.new(@username, @password, @api_key)
      assert_equal '/base/feeds/snippets/', another_client.feed_path
    end

    def test_update_item

      base_id = @base_client.create_item(@item)
      assert base_id.kind_of?(String)
      @item.add_label('updated_label')
      assert @base_client.update_item(@item)
    end

    def test_post_simple_item
      
      base_id = @base_client.create_item(@item)
      assert base_id.kind_of?(String)
    end

     def test_delete_item
       
       assert @base_client.create_item(@item)
       assert @base_client.delete_item(@item)
     end

    # sanity test to assert the network is up
    def test_getter

      resp = Net::HTTP.get_response('www.google.com', '/base/feeds/itemtypes/en_GB/jobs')   
      assert resp.kind_of?(Net::HTTPOK)
    end


  end

end



lib/base4r.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007, 2008 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#
# Require files we depend on
# 
#
require 'base_client'
require 'item'
require 'attribute'


lib/attribute.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007, 2008 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#
# following taken from http://code.google.com/apis/base/starting-out.html
# 
# There are some limitations on how long your attribute names and values can be and how many you can use:

#     * titles: between 3-1000 characters
#     * attributes:
#           o names: 30 characters
#           o attribute text: 1000 characters, including spaces
#           o total number permitted: 30 per item
#     * labels:
#           o names: 40 characters
#           o total number permitted: 10 per item
#     * URL length: 1000

module Base4R

  # Attributes are typed key-value pairs that describe content. Attributes describe the
  # Base Item. Client code should use subclasses of Attribute such as TextAttribute, 
  # BareAttribute, DateTimeAttribute, IntAttribute, FloatUnitAttribute, UrlAttribute, 
  # LocationAttribute, BooleanAttribute. 
  # Each of these can represent themselves as required by the Atom format.
  class Attribute

    attr_accessor :name
    attr_accessor :value
    attr_accessor :type_name
    attr_accessor :namespace
    
    # Represent this Attribute as an XML element that is a child of _parent_.
    def to_xml(parent)

      el = parent.add_element(calc_el_name())

      if @namespace == :g then
        el.add_attribute('type', @type_name)
      end
      el.text=@value
    end
    
    protected

    def initialize(name, value, namespace)
      @name = name
      @value = value
      @namespace = namespace
    end

    def calc_el_name
      if @namespace == :g then
        el_name = 'g:'+@name
      else 
        el_name = @name
      end
      el_name
    end
 
 end

  # TextAttribute is a simple string.
  class TextAttribute < Attribute
    
    # Create a TextAttribute with _name_, _value_ and in _namespace_
    def initialize(name, value, namespace)
      @type_name = 'text'
      super(name, value, namespace)
    end
  end

  # BareAttribute is a string attribute but is in the google namespace
  class BareAttribute < Attribute
    
    # Create a BareAttribute with _name_ and _value_
    def initialize(name, value)
      super(name, value, :g)
    end
    
    def to_xml(parent)
      el = parent.add_element('g:'+name).text=value
    end
  end

  # UrlAttribute represents a URL
  class UrlAttribute < Attribute
    
    # Create a UrlAttribute with _name_, _value_ and _namespace_
    def initialize(name, value, namespace)
      @type_name = 'url'
      super(name, value, namespace)
    end
    
    def to_xml(parent)
     
      urlel = parent.add_element(calc_el_name())
      urlel.add_attribute('rel', 'alternate')
      urlel.add_attribute('type', 'text/html')
      urlel.add_attribute('href', @value)
    end

  end

  # DateTimeAttribute represents a DateTime
  class DateTimeAttribute < Attribute

    def initialize(name, value, namespace)
      @type_name = 'dateTime'
      super(name, value, namespace)
    end
  end

  # IntAttribute represents an Integer
  class IntAttribute < Attribute

    def initialize(name, value, namespace)
      @type_name = 'int'
      super(name, value, namespace)
    end
  end

  # BooleanAttribute represents a Boolean
  class BooleanAttribute < Attribute

    def initialize(name, value, namespace)
      @type_name = 'boolean'
      super(name, value, namespace)
    end
  end

  # FloatUnitAttribute represents a floating-point property with units.
  class FloatUnitAttribute < Attribute

    # Create a FloatUnitAttribute with _name_, a quantity of _number_, described in _units_ and in _namespace_ 
    def initialize(name, number, units, namespace)
      @type_name = 'floatUnit'
      super(name, number.to_s+' '+units, namespace)
    end
  end

  # LocationAttribute represents an Item's location
  class LocationAttribute < Attribute

    # Create a LocationAttribute with _name_ and in _location_
    def initialize(name, location)
      @type_name = 'location'
      super(name, location, :g)
    end
  end
  
end


lib/base_client.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007, 2008 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

require 'net/http'
require 'client_login'

module Base4R

  # BaseClient handles all communication with the Base API using HTTP
  class BaseClient

    ITEMS_PATH = '/base/feeds/items/'
    SNIPPETS_PATH = '/base/feeds/snippets/'    
    BASE_HOST = 'base.google.com'

    attr_reader :auth_key #:nodoc:
    attr_reader :feed_path #:nodoc:

    # Construct a BaseClient, which will make API requiest for the Base account
    # belonging to _username_, authenticating with _password_ and using _api_key_. 
    # Requests will be made against the public feed if _public_feed_ is true, which is the default.
    # The BaseClient can be used for a number of Base API requests.
    # 
    def initialize(username, password, api_key, public_feed=true)

      @auth_key = ClientLogin.new.authenticate(username, password)
      @api_key = api_key

      if public_feed then
        @feed_path = SNIPPETS_PATH
      else
        @feed_path = ITEMS_PATH
      end
    end

    # Creates the supplied _item_ as a new Base Item.
    # Throws an Exception if there is a problem creating _item_.
    #
    def create_item(item)
      
      resp = do_request(item.to_xml.to_s, 'POST')
      raise "Error creating base item:"+resp.body unless resp.kind_of? Net::HTTPCreated
      resp['location'] =~ /(\d+)$/
      item.base_id= $1
      $1
    end

    # Update the supplied Base _item_. Returns true on success.
    # Throws an Exception if there is a problem updating _item_.
    #
    def update_item(item)

      raise "base_id is required" if item.base_id.nil?
      resp = do_request(item.to_xml.to_s, 'PUT', item.base_id)
      raise "Error updating base item:"+resp.body unless resp.kind_of? Net::HTTPOK
      true
    end

    # Delete the supplied Base _item_. Returns true on success.
    # Throws an Exception if there is a problem deleting _item_
    #
    def delete_item(item)

      raise "base_id is required" if item.base_id.nil?
      resp = do_request(nil, 'DELETE', item.base_id)
      raise "Error deleting base item:"+resp.body unless resp.kind_of? Net::HTTPOK
      true
    end

    private

    def do_request(data, http_method, base_id=nil)

      url = 'http://'+BASE_HOST+@feed_path
      url << base_id unless base_id.nil?
      
      url = URI.parse(url)

      headers = {'X-Google-Key' => 'key='+@api_key,
                 'Authorization' => 'GoogleLogin auth='+@auth_key,
                 'Content-Type' => 'application/atom+xml'}

      result = Net::HTTP.start(url.host, url.port) { |http|
        http.send_request(http_method, url.path, data, headers)
      }
      
    end

  end 
end


lib/client_login.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007, 2008 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#
# Authenticates the user with google ClientLogin
#
#
require 'net/https'

module Base4R

  #
  # ClientLogin authenticates the user with Google Accounts using the ClientLogin service.
  #
  class ClientLogin

    ServiceURL = URI.parse('https://www.google.com/accounts/ClientLogin')

    # Create a ClientLogin. _options_ is optional and by default logs into
    # the Google Base API.
    def initialize(options={})

      defaults = {:accountType => 'GOOGLE',
                  :service => 'gbase',
                  :source => 'base4r-clientloginapp-v1'}

       @config = defaults.merge(options)
    end

    # attempt to authenticate _email_ and _password_ with Google.
    # Returns an authentication token on success, raises Exception on failure.
    # todo: make exception handling more useful
    def authenticate(email, password)

      params = {
        'accountType' => @config[:accountType],
        'Email' => email,
        'Passwd' => password,
        'service' => @config[:service],
        'source' => @config[:source]
      }
      
      req = Net::HTTP::Post.new(ServiceURL.path)
      req.set_form_data(params)

      http = Net::HTTP.new(ServiceURL.host, 443)
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
      http.ca_file = File.expand_path(File.dirname(__FILE__)+"/../cert/cacert.pem")

      # todo - above is probably reading file for every request, cache certs instead

      resp = http.request(req)
            
      unless resp.instance_of? Net::HTTPOK then
        resp.body =~ /^Error=(.+)$/

        raise CaptchaRequiredException.new(email) if 'CaptchaRequired' == $1        
        raise 'unknown exception authenticating with google:'+$1
      end

      # parse the auth key from the body
      resp.body =~ /^Auth=(.+)$/
      return $1

    end
  end

  # thrown when ClientLogin fails to authenticate because the ClientLogin API has 
  # requested that the user complete a captcha validation. 
  class CaptchaRequiredException < Exception
    
    attr_reader :email
    
    # Create a CaptchaRequiredException to notify the user that _email_ must complete
    # a captcha validation
    def initialize(email)
      @email = email
    end
  end

end


lib/item.rb
# Base4R is a ruby interface to Google Base
# Copyright 2007, 2008 Dan Dukeson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

require 'rexml/document'

module Base4R

  # Item describes a single entry that will be or is already stored in Google Base
  class Item

    # array of Attribute objects that describe the Item
    attr_accessor :attributes
    
    # Name of the creator of the Item
    attr_accessor :author_name
    
    # Email of the creator of the Item
    attr_accessor :author_email
    
    # Title of the Item
    attr_accessor :title

    # ID of this Item as assigned by Google
    attr_accessor :base_id

    # unique alphnumeric identifier for the item - e.g. your internal ID code. 
    # IMPORTANT: Once you submit an item with a unique id, this identifier must 
    # not change when you send in a new data feed. Each item must retain the same 
    # id in subsequent feeds.
    attr_accessor :unique_id

    # Represents this Item as Atom XML which is the format required by the Base API. 
    def to_xml
      
      doc = REXML::Document.new('<?xml version="1.0" ?>')

      entry = doc.add_element('entry', {
                                'xmlns'=>'http://www.w3.org/2005/Atom', 
                                'xmlns:g' => 'http://base.google.com/ns/1.0'
                              })
      
      author = entry.add_element('author')
      author.add_element('name').text= @author_name
      author.add_element('email').text= @author_email

      entry.add_element('category',
                        {'scheme'=>'http://www.google.com/type',
                         'term' => 'googlebase.item'
                        }
                       )

      entry.add_element('title').text= @title
      
      @attributes.each do |attr|
        attr.to_xml(entry)
      end

      doc
    end

  end

  #
  # Item with a minimal set of Attributes, extend for specific Item Types
  # 
  #
  class UniversalItem < Item

    # Create a new UniversalItem, with _unique_id_, created by _author_name_ who's email is _author_email_, 
    # described by _description_, found at URL _link_, entitled _title_, phone number is 
    # _contact_phone_, item type is _item_type_, _target_country_ e.g. 'GB', _item_language_ e.g. 'EN' 
    def initialize(unique_id, author_name, author_email, description, link, 
                   title, contact_phone, item_type, target_country, item_lang)
     
      @author_name  = author_name
      @author_email = author_email
      @title = title
      
      @attributes = []

      @attributes << UrlAttribute.new('link', link, :atom)
      @attributes << TextAttribute.new('content', description, :atom)
      @attributes << TextAttribute.new('contact_phone', contact_phone, :g)
      @attributes << TextAttribute.new('item_language', item_lang, :g)
      @attributes << TextAttribute.new('item_type', item_type, :g)
      @attributes << TextAttribute.new('target_country', target_country, :g)
      @attributes << TextAttribute.new('id', unique_id, :g)
    end

    def application=(application)
      @attributes << TextAttribute.new('application', application, :g)
    end
    
    def expiration_date=(expires)
      @attributes << DateTimeAttribute.new('expiration_date', expires, :g)
    end

    def add_image_link(image_url)
      @attributes << BareAttribute.new('image_link', image_url)
    end

    def add_label(label)
      @attributes << TextAttribute.new('label', label, :g)
    end


  end

  # ProductItem is a standard item type. This class includes some of the attributes that are
  # suggested or required for the Product item type.
  #
  class ProductItem < UniversalItem

    def initialize(unique_id, author_name, author_email, description, link, 
                   title, contact_phone, item_type, target_country, item_lang)

      super(unique_id, author_name, author_email, description, link,
            title, contact_phone, item_type, target_country, item_lang)
      
    end
   

    def condition=(condition)
      @attributes << TextAttribute.new('condition', condition, :g)
    end

    def delivery=(will_deliver)

      @attributes << BooleanAttribute.new('delivery', will_deliver.to_s, :g)
    end

    def delivery_notes=(notes)
      
      @attributes << TextAttribute.new('delivery_notes', notes, :g)
    end

    def department=(department)
      
      @attributes << TextAttribute.new('department', department, :g)
    end

    def add_payment=(payment)
      
      @attributes << TextAttribute.new('payment', payment, :g)
    end

    def payment_notes=(notes)
      
      @attributes << TextAttribute.new('payment_notes', notes, :g)
    end

    def pickup=(pickup)
      
      @attributes << BooleanAttribute.new('pickup', false, :g)
    end

    def price(price_amount, price_units)
      
      @attributes << FloatUnitAttribute.new('price', price_amount, price_units, :g)
    end

    def price_type=(price_type)
      
      @attributes << TextAttribute.new('price_type', price_type, :g)
    end

    def price_units=(price_units)
      
      @attributes << TextAttribute.new('price_units', price_units, :g)
    end

    def location=(loc)

      @attributes << LocationAttribute.new('location', loc)
    end

    def product_type=(prod_type)
      
      @attributes << TextAttribute.new('product_type', prod_type, :g)
    end

    def add_custom(name, value)

      if name.downcase == 'quantity' then
        @attributes << IntAttribute.new('Quantity', value, :g)
        return
      end

      self.add_custom_text(name, value)
    end
    

    def add_custom_text(name, value)

      @attributes << TextAttribute.new(name, value, :g)
    end


  end
  
end
